### PR TITLE
Enhancements to the Calcite test framework

### PR DESCRIPTION
**Pull request by paul-rogers**
Follow-up to the earlier Calcite test framework PR. This version makes more of the components customizable, and provides additional options for the test builder. The need for these items, and verification of their functionality, comes from tests for a few private extensions.

This PR also standardizes various "Unauthorized" messages to avoid the persistent issues with security ITs where the expected message does not match the one Druid delivers, because Druid used multiple variations of the same message.

#### Release note

There are no user-visible changes in the PR. All changes relate to unit tests.

<hr>

This PR has:

- [X] been self-reviewed.
- [X] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [X] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [X] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.

### CODE REVIEW

**Review by kfaraz — APPROVED** [review]
LGTM

File: sql/src/test/java/org/apache/druid/sql/calcite/QueryTestBuilder.java [hunk 1]
```diff
@@ -200,13 +202,46 @@ public QueryTestBuilder authConfig(AuthConfig authConfig)
...
+  protected PlannerFixture plannerFixture()
+  {
+    if (plannerFixture != null) {
+      return plannerFixture;
+    } else {
+      return config.plannerFixture(plannerConfig, authConfig);
+    }
+  }
+
   public void run()
```
Comment (rohangarg) [inline sql/src/test/java/org/apache/druid/sql/calcite/QueryTestBuilder.java:238]: is there any reason to have the run and results methods in `runner` and `builder` both? also, should the builder be called `QueryTestRunnerBuilder`?

File: sql/src/test/java/org/apache/druid/sql/calcite/util/SqlTestFramework.java [hunk 2]
```diff
@@ -127,6 +142,14 @@ SqlEngine createEngine(
     void configureJsonMapper(ObjectMapper mapper);
 
     void configureGuice(DruidInjectorBuilder builder);
+
+    Set<ExtensionCalciteRuleProvider> calciteRules();
```
Comment (rohangarg) [inline sql/src/test/java/org/apache/druid/sql/calcite/util/SqlTestFramework.java:146]: could be `extensionCalciteRules`

File: sql/src/test/java/org/apache/druid/sql/calcite/util/SqlTestFramework.java [hunk 3]
```diff
@@ -127,6 +142,14 @@ SqlEngine createEngine(
...
+
+    ViewManager createViewManager();
+
+    void populateViews(ViewManager viewManager, PlannerFactory plannerFactory);
```
Comment (rohangarg) [inline sql/src/test/java/org/apache/druid/sql/calcite/util/SqlTestFramework.java:150]: should this be a part of the `SqlQueryFramework` instead?

File: sql/src/test/java/org/apache/druid/sql/calcite/util/SqlTestFramework.java [hunk 4]
```diff
@@ -240,50 +348,107 @@ public SqlTestFramework build()
...
+      );
+      this.statementFactory = QueryFrameworkUtils.createSqlStatementFactory(
+          framework.engine,
+          plannerFactory,
+          authConfig
+      );
+      componentSupplier.populateViews(viewManager, plannerFactory);
+    }
+
+    public ViewManager viewManager()
```
Comment (rohangarg) [inline sql/src/test/java/org/apache/druid/sql/calcite/util/SqlTestFramework.java:400]: where would these get used? incase they are not used, can we call this `SqlStatementFactoryFixture`?

File: sql/src/test/java/org/apache/druid/sql/calcite/util/SqlTestFramework.java [hunk 5]
```diff
@@ -349,83 +514,12 @@ public QueryRunnerFactoryConglomerate conglomerate()
    * factory is specific to one test and one planner config. This method can be
    * overridden to control the objects passed to the factory.
    */
-  public SqlStatementFactory statementFactory(
+  public PlannerFixture plannerFixture(
```
Comment (rohangarg) [inline sql/src/test/java/org/apache/druid/sql/calcite/util/SqlTestFramework.java:517]: the documentation seems a bit more related to the old method

**Review by rohangarg — COMMENTED** [review]
Thanks a lot for the changes @paul-rogers! Left some comments

**Pull request merged**